tests/core/editable/insertion.js

```diff
@@ -159,7 +159,79 @@
 			editor.insertHtml( '<p>bam</p>', 'html', range );
 
 			assert.isInnerHtmlMatching( '<p id="p1">foo</p><p>bam^@</p><p>bar@</p>',
-				bender.tools.selection.getWithHtml( editor ), { compareSelection: true, normalizeSelection: true } );
+				tools.selection.getWithHtml( editor ), { compareSelection: true, normalizeSelection: true } );
 		}
-	} );
+	};
+
+	// Add tests for (#2813).
+	addInsertionTests( [
+		{
+			name: 'test insert when selection starts at the beginning of span',
+			initial: '<span>{foo}bar</span>',
+			data: '<div>div</div>',
+			expected: '<div>div</div><span>bar</span>'
+		}, {
+			name: 'test insert when selection ends at the end of span',
+			initial: '<span>foo{bar}</span>',
+			data: '<div>div</div>',
+			expected: '<span>foo</span><div>div</div>'
+		}, {
+			name: 'test insert when selection covers span',
+			initial: '<span>{foobar}</span>',
+			data: '<div>div</div>',
+			expected: '<div>div</div>'
+		}, {
+			name: 'test insert when selection is in the middle of span',
+			initial: '<span>foo{bar}&nbsp;</span>',
+			data: '<div>div</div>',
+			expected: '<span>foo</span><div>div</div><span>&nbsp;</span>'
+		}
+	] );
+
+	bender.test( tests );
+
+	function addInsertionTests( tests ) {
+		CKEDITOR.tools.array.forEach( tests, addInsertionTest );
+	}
+
+	function addInsertionTest( options ) {
+		var methods = [ 'insertHtml', 'insertHtmlIntoRange', 'insertElement', 'insertElementIntoRange', 'insertText' ];
+
+		CKEDITOR.tools.array.forEach( methods, function( methodName ) {
+			tests[ options.name.replace( 'insert', methodName ) ] = function() {
+				var editor = this.editorBot.editor;
+
+				tools.selection.setWithHtml( editor, options.initial );
+
+				if ( methodName === 'insertText' ) {
+					assertInsertText( editor, options.initial.replace( /{.*}/, 'text' ), 'text' );
+				} else {
+					assertInsertionMethod( editor, options.expected, methodName, options.data  );
+				}
+			};
+		} );
+	}
+
+	function assertInsertionMethod( editor, expected, methodName, html ) {
+		var useRange = methodName.indexOf( 'Range' ) >= 0,
```

Comment:
Maybe you could make it even more readable by using a `switch` statement instead of string sniffing directly from `addInsertionTests`?